hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/PublishingHandler.java

```diff
@@ -0,0 +1,220 @@
+    private final MessageReader messageReader;
+    private final int defaultAsyncTimeout;
+    private final int longAsyncTimeout;
+    private final boolean previewEnabled;
+
+    @Inject
+    public PublishingHandler(TopicsCache topicsCache, MessageFactory messageFactory, BrokerMessageProducer brokerMessageProducer,
+                             MessagePreviewLog messagePreviewLog, MessageErrorProcessor messageErrorProcessor,
+                             MessageEndProcessor messageEndProcessor, ConfigFactory configFactory) {
+
+        this.topicsCache = topicsCache;
+        this.messageFactory = messageFactory;
+        this.brokerMessageProducer = brokerMessageProducer;
+        this.messagePreviewLog = messagePreviewLog;
+        this.messageErrorProcessor = messageErrorProcessor;
+        this.messageEndProcessor = messageEndProcessor;
+        this.messageReader = new MessageReader();
+        this.defaultAsyncTimeout = configFactory.getIntProperty(Configs.FRONTEND_IDLE_TIMEOUT);
+        this.longAsyncTimeout = configFactory.getIntProperty(Configs.FRONTEND_LONG_IDLE_TIMEOUT);
+        this.previewEnabled = configFactory.getBooleanProperty(FRONTEND_MESSAGE_PREVIEW_ENABLED);
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        if (exchange.isInIoThread()) {
+            // switch to worker thread
+            exchange.dispatch(this);
+            return;
+        }
+
+        onTopicPresent(exchange, topicWithMetrics -> {
+            exchange.addExchangeCompleteListener(new ExchangeMetrics(topicWithMetrics));
+            read(exchange, topicWithMetrics);
+        });
+    }
+
+    private void read(HttpServerExchange exchange, TopicWithMetrics topicWithMetrics) {
+        MessageState state = new MessageState();
+        int timeout = topicWithMetrics.getTopic().isReplicationConfirmRequired() ? longAsyncTimeout : defaultAsyncTimeout;
+
+        MessageTimeout messageTimeout = new MessageTimeout(state.messageId());
+        messageTimeout.setTimeoutKey(exchange.getIoThread().executeAfter(
+                () -> exchange.dispatch(() -> handleTimeout(exchange, state, topicWithMetrics.getTopic(), messageTimeout, timeout)),
+                timeout,
+                MILLISECONDS));
+
+        messageReader.onMessageRead(exchange, state, topicWithMetrics, messageTimeout,
+                messageContent -> processAndLogPreview(exchange, topicWithMetrics, state, messageTimeout, messageContent),
+                exception -> {
+                    messageTimeout.remove();
+                    if (exception instanceof InvalidContentLengthException) {
+                        messageErrorProcessor.invalidLength(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
+                    } else {
+                        messageErrorProcessor.readException(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
```

Comment:
This looks very scary to me. Any exception that is not ```InvalidContentLengthException``` will leave response without a proper response code cause ```messageErrorProcessor.invalidLength``` handles responses and generic ```readException``` doesn't.